Std/Data/List/Basic.lean

```diff
@@ -13,7 +13,37 @@ namespace List
 
 /-! ## Tail recursive implementations for definitions from core -/
 
-/-- Tail recursive version of `erase`. -/
+/-- Tail recursive version of `List.any`, with short-circuiting. -/
+def anyTR (xs : List α) (p : α → Bool) : Bool :=
+  go false xs
+where
+  /-- Auxiliary for `anyTR`. -/
+  go : Bool → List α → Bool
+  | true, _ => true
+  | false, [] => false
+  | false, h :: t => go (p h) t
```

Comment:
How about the following?

```suggestion
def anyTR : List α -> (α → Bool) -> Bool
| [], _ => false
| h :: t, p => p h || anyTR t p
```

Reply:
After inspecting the C code in stage0, it seems that `List.or`, which is the specialization of `List.any` where `p=id`, compiles as short-circuiting tail-recursive code. I'm not sure about `List.any` though since it has a lot more C crud.

I wonder if the short-circuiting tail-recursive version should actually be in core?

Reply:
💡 Interesting. I wish we had an annotation to require / check tail-recursiveness...

Reply:
@fgdorais my [tail recursion detector](https://leanprover.zulipchat.com/#narrow/stream/348111-std4/topic/tail.20recursion/near/404515501) says that `List.any`/`List.all` is not tail-recursive, but `List.or`/`List.and` is, which is a bit mysterious.

Reply:
I'm ok with merging `anyTR`/`allTR` if that would unblock something, but making the change to core seems preferable.